src/plugins/data/common/index_patterns/index_patterns/ensure_default_index_pattern.ts

```diff
@@ -44,26 +42,33 @@ export const createEnsureDefaultIndexPattern = (
    * one otherwise.
    */
   return async function ensureDefaultIndexPattern(this: IndexPatternsContract) {
+    const datasources = await savedObjectsClient.find({ type: 'data-source' });
+    const indexPatterns = await savedObjectsClient.find({ type: 'index-pattern' });
+    const existDataSources = datasources.map((item) => item.id);
+    const availableArray: string[] = [];
+    indexPatterns.forEach((item) => {
+      const refId = item.references[0]?.id;
+      const refIdBool = !!refId;
+      if (!refIdBool || existDataSources.includes(refId)) {
+        availableArray.push(item.id);
+      }
+    });
     if (canUpdateUiSetting === false) {
       return;
     }
-    const patterns = await this.getIds();
     let defaultId = await uiSettings.get('defaultIndex');
```

Comment:
What about optimizing the check flow to:

1. Get the data source from references of default index pattern
2. If there is no data source, we are good, if there is data source, try to get the data source info by using savedObjectClient.get
3. If the call fails(the data source deleted / dissociated from the workspace), get index patterns / data sources list to see if there is other index patterns can be a good candidate of default index pattern
4. Set the one as default index pattern.

By doing so, no additional call is required if the index pattern and its underlying data source is good.